### DISCUSSION TITLE
Newer version is not being detected but the Delta applies when done manualy.

### DISCUSSION THREAD

**Discussion by nishaanth-ms-21155** [sparkle-project/Sparkle#2635] [Q&A]
Using Sparkle 1.24 (ik it's old but it worked till now)

I have created two MyApp.app files v1234.13 and v1300.46 I have created the deltas between them, sgned it properly and BinaryDelta apply works. The problem I encounter with appcast looks like this.

I serve my appcast and new zip of MyApp.app v1300.46 and set my feed url to localhost and the appcast something looks like this:

````
<?xml version="1.0" encoding="utf-8"?>
<rss version="2.0" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:sparkle="http://www.andymatuschak.org/xml-namespaces/sparkle">
    <channel>
        <title>MyApp</title>
        <link>http://localhost:8000</link>
        <description></description>
        <language>en</language>
        <item>
            <title>MyApp 1.300.46</title>
            <description>Beta Release - Major 2, Minor 24, Patch 0</description>
            <pubDate>Tue, 24 Sep 24 19:00:00 +0530</pubDate>
            <enclosure sparkle:shortVersionString="1.300.46" sparkle:version="1300.46" url="MyApp.app " />
            <sparkle:deltas>
                <enclosure url="http://localhost:8000/MyApp-v1.234.13-v1.300.46.delta"
                    type="application/octet-stream"
                    sparkle:version="1.300.46"
                    sparkle:shortVersionString="1.300.46"
                    sparkle:deltaFrom="1234.13"
                    sparkle:edSignature="<sign>"
                    length="53172706"
                />
            </sparkle:deltas>
        </item>
    </channel>
</rss>
```
For some reason if I change just the appcast data like 1301.46 as latest, the given data applies through appcastas expected but no lesser than 1300.46 even though the current app is 1234.13

**Comment by zorgiepoo** [comment]
Please don't pass `sparkle:version` and `sparkle:shortVersionString` in the enclosure items. Pass them as top level elements instead and remove the duplication redundancy. See the reference appcast here -- https://sparkle-project.org/documentation/publishing/#update-your-appcast

One obvious issue at least is the `sparkle:version` you passed for the delta item enclosure (`1.300.46`) is different than the one in the main parent enclosure (`1300.46`). The version in your delta item enclosure is wrong.

You can also use generate_appcast to automatically generate appcasts instead of manual generation.

**Reply by nishaanth-ms-21155** [reply]
Thanks for the help. I have matched the CFBundle version and ShortString version appropriately till now in Info.plist and it has worked till now. I have tried it in another PC which applies the delta correctly as expected. I am guesssing the issue is that a day prior, I was randomly prompted a window from Sparkle that an update is available where the updateCheck network call is sent. This was never implemented and needs to be disabled. Anyways, I clicked, "Skip this Version" which i am guessing ignores all updates below the version I skipped which is why changing the appcast latest version value updates correctly and works on another machine because i didn't click it there. Any help on that side? 
My questions are: Why does the window appear and how to disable it? If someone clicks Skip this version, is there no way to update to that version till the next update comes?

**Reply by nishaanth-ms-21155** [reply]
Update: Tried Removing Sparkle:version and shortStrignVersion in delta enclosure and myApp crashes.

**Reply by zorgiepoo** [reply]
> Thanks for the help. I have matched the CFBundle version and ShortString version appropriately till now in Info.plist and it has worked till now. I have tried it in another PC which applies the delta correctly as expected.

You didn't seem to acknowledge the sparkle:version you passed to your delta item enclosure you pasted here is wrong (i.e. it should be 1300.46, not 1.300.46)

>  I was randomly prompted a window from Sparkle that an update is available where the updateCheck network call is sent. This was never implemented and needs to be disabled.

By default Sparkle handles doing periodic updates checks if the user gives permission for Sparkle to do this, see documentation for SUEnableAutomaticChecks in https://sparkle-project.org/documentation/customization/

> Anyways, I clicked, "Skip this Version" which i am guessing ignores all updates below the version I skipped

Yes that is how the feature works. If a newer version is released or if the user manually checks for updates, this resets.

> Why does the window appear and how to disable it?

I already explained this here https://github.com/sparkle-project/Sparkle/discussions/2623

> Update: Tried Removing Sparkle:version and shortStrignVersion in delta enclosure and myApp crashes.

Look at the crash log from Console.app and attach it here if it's in Sparkle's code to see why this happened. You also need to add sparkle:verson and sparkle:shortVersionString elements in your main item like this:

```
<item>
    <title>MyApp 1.300.46</title>
    <description>Beta Release - Major 2, Minor 24, Patch 0</description>
    <pubDate>Tue, 24 Sep 24 19:00:00 +0530</pubDate>
    <sparkle:version>1300.46</sparkle:version>
    <sparkle:shortVersionString>1.300.46</sparkle:shortVersionString>
    <enclosure url="MyApp.app" sparkle:edSignature="<sign>" />
    <sparkle:deltas>
        <enclosure url="http://localhost:8000/MyApp-v1.234.13-v1.300.46.delta"
            type="application/octet-stream"
            sparkle:deltaFrom="1234.13"
            sparkle:edSignature="<sign>"
            length="53172706"
        />
    </sparkle:deltas>
</item>
```

**Comment by nishaanth-ms-21155** [comment]
> You didn't seem to acknowledge the sparkle:version you passed to your delta item enclosure you pasted here is wrong (i.e. it should be 1300.46, not 1.300.46)
Mistake on pasting in github. Kindly ignore.

> I already explained this here https://github.com/sparkle-project/Sparkle/discussions/2623

Makes sense. Have minimized the UI where it asks for an update to user. by setting config in updater object

> Yes that is how the feature works. If a newer version is released or if the user manually checks for updates, this resets.

I intend to provide any info regarding updates within the app a separate window doesn't makes sense in my UX since all things happen within the app. Does skipping a version store data in the user machine as if the app is updated to the latest version? Is there anyway to revert that, atleast on the developer side of things, preferably user side? 

> Look at the crash log from Console.app and attach it here if it's in Sparkle's code to see why this happened. 
I have tied it yet. Can you guide me with any docs/discussion?


> You also need to add sparkle:verson and sparkle:shortVersionString elements in your main item like this:

Awesome. Will try that.

Thanks for the reply and help.

**Reply by zorgiepoo** [reply]
> Does skipping a version store data in the user machine as if the app is updated to the latest version? Is there anyway to revert that, atleast on the developer side of things, preferably user side?

Sparkle stores it in the NSUserDefaults of the app, I think using the `SUSkippedVersion` key.

> I intend to provide any info regarding updates within the app a separate window doesn't makes sense in my UX since all things happen within the app.

As far as UI customizability goes, Sparkle 2 has an API for custom user interfaces
https://sparkle-project.org/documentation/api-reference/Protocols/SPUUserDriver.html

and also an API for gentle reminders on top of the standard UI
https://sparkle-project.org/documentation/gentle-reminders/

Some UI is inevitable for users that need to perform authenticated installs.

**Reply by nishaanth-ms-21155** [reply]
Authenticated installs can be limited to Access prompts to directories and admin logins? Why provide a window for user action which basically asks for confirmation? 

Can I know in which of the cases throughout Sparkle does the UpdateAlert Window appear?

> Sparkle stores it in the NSUserDefaults of the app, I think using the SUSkippedVersion key.

I tested it and noticed it stores thoughout a apple user in a machine. Is there a workaround to perform manual update checks inspite of user choosing "Skip this Version" Or any dev side workaround to delete `NSUserDefaults` ?

**Reply by zorgiepoo** [reply]
Authentication is needed if Sparkle has insufficient permission to replace the app as the non-root user (i.e, if the owner/group of the app bundle or its parent directory differs from the process). Run `sudo chown root path-to-your-app-bundle` and try updating it to simulate this. A typical example is a standard/non-admin user trying to update your app in /Applications, but there are other cases where this happens and while it's not the majority of users it's common enough you'll have users run into this.

As I said before and again, for the non-authenticated case, enable automatic downloading/installing of updates ([SUAutomaticallyUpdate](https://sparkle-project.org/documentation/customization/)). `SUEnableAutomaticChecks` may also be of interest there. There is [another delegate method](https://sparkle-project.org/documentation/api-reference/Protocols/SPUUpdaterDelegate.html#/c:objc(pl)SPUUpdaterDelegate(im)updater:willInstallUpdateOnQuit:immediateInstallationBlock:) if you want to intercept the delayed 1-week timer / reminder.

Ignoring the user's request for skipping updates is a bad idea. Better to focus on making the update process automatic as much as possible, and still present UI in cases where you have no choice, or want users to perform checks themselves manually. If you want to replace Sparkle's UI with your own, see my above links on UI customizability.

Some of what I say may pertain to Sparkle 2 only. I'm not providing support for older versions at this kind of extent.

**Reply by zorgiepoo** [reply]
> I tested it and noticed it stores thoughout a apple user in a machine. Is there a workaround to perform manual update checks inspite of user choosing "Skip this Version" Or any dev side workaround to delete NSUserDefaults ?

Developer workaround:
```
defaults delete my-app-bundle-id SUSkippedVersion
```

A manual update check from the user is performed through checkForUpdates. This is what you would call if you have a menu item to check for updates, or a button the user clicks to check for updates, etc. This also resets the skipped version state and shows the updater UI.

**Reply by nishaanth-ms-21155** [reply]
> A manual update check from the user is performed through checkForUpdates. 

`checkForUpdatesinBackground` is what I am using. App is built all around AutomaticChecks but the idea is to remove all pop up windows and provide update information within the app. User triggered checks and corner notifications within app is provided to notify updates. So the popup to update doesnt make sense to me and is not expected especially when it is a confirmation and not an authorization to overwrite /Applications content. Users request to skip current updates is fine until they noticed there actually is an update and want to update now since app releases are 1 month away and users are notified through a second source also. The second source and sparkle mismatch update availability since Sparkle controls all the updates yet it shows it is upto date since they skipped and wouldn't let it update. Hence, remove all pop ups uness it requires a password.

**Reply by zorgiepoo** [reply]
From Sparkle's point of view:

* `checkForUpdatesInBackground` is wrong for user-triggered update checks. [checkForUpdates](https://sparkle-project.org/documentation/api-reference/Classes/SPUUpdater.html#/c:objc(cs)SPUUpdater(im)checkForUpdatesInBackground) is intended for user-initiated checks. That's partially why you're fighting skipped updates. checkForUpdatesInBackground is intended for background update checks that the user does not trigger.
* Sparkle only supports replacing its UI (including you see presented with user initiated checks) with your own UI using [SPUUserDriver](https://sparkle-project.org/documentation/api-reference/Protocols/SPUUserDriver.html) (Sparkle 2). Few apps with specific UI needs use this. You could even not provide an option to skip updates, and use your own window/whatever. (Note this is not a way to hide or show no UI; this API assumes you will show UI and the user will make explicit choices).
* Sparkle can automatically download and install updates in the background silently if configured to do so, without showing any UI (implying not even showing your custom UI), in some but not all cases. This mode is intended for showing no UI when possible. It's not a way intended to replace Sparkle's UI and show your own UI. (If an update is being installed automatically in the background, the user can still trigger a user-initiated update check to bring it in the foreground, or UI may be presented if authorization is required after the update has been downloaded, using the SPUUserDriver).

**Reply by nishaanth-ms-21155** [reply]
Nice. That' all I needed. I switched to an own implementation by calling `checkForUpdatesWithDriver` passing `SUAutomaticUpdateDriver` so no relation to `SUUIBasedUpdateDriver` exists since I found that `SUScheduledUpdateDriver` inherits  `SUUIBasedUpdateDriver`. That solved 99 percent of my problems. I am just worried about Scheduled Automatic checks where `SUScheduledCheckInterval` existing itself worries me. After this forked change is there any possibility of UI appearing with /Applications access, scheding checks or whatever else which might show UI? Or is it better to remove the class entirely in the fork?
Thanks.

**Reply by zorgiepoo** [reply]
That is not showing understanding of my summary.. I do not support any of that customization as it's incompatible with modern versions of Sparkle (those are private implementation details of Sparkle 1). Obviously you will need the UI based update driver for authorized installs otherwise you will have a buggy UX. What you're trying to fork should be unnecessary if you just opt into automatic downloads/installs ([read SUAutomaticallyUpdate docs](https://sparkle-project.org/documentation/customization/)) which prefers using the automatic update driver when feasible in the first place.

You're trying to hack/fork an outdated, vulnerable and slower and less extensible, version of Sparkle rather than upgrade and using the newer properly supported APIs for UI extensibility. I will not help you there, and it's a waste of time.

**Reply by nishaanth-ms-21155** [reply]
Yeah, completely forgot about the password. Thanks for all the help. Will try to move to Sparkle 2 soon enough.
